contracts/paymaster/PaymasterTopFacet.sol

```diff
@@ -0,0 +1,173 @@
+  bool public useRejectOnRecipientRevert = false;
+
+  // These parameters are documented in IPaymaster.GasAndDataLimits.
+  // SEE: https://github.com/opengsn/gsn/blob/v3.0.0-beta.10/packages/contracts/src/interfaces/IPaymaster.sol#L29
+
+  // Overhead of forwarder verify+signature, plus hub overhead.
+  uint256 public constant FORWARDER_HUB_OVERHEAD = 50000;
+  uint256 public constant PRE_RELAYED_CALL_GAS_LIMIT = 100000;
+  uint256 public constant POST_RELAYED_CALL_GAS_LIMIT = 110000;
+  uint256 public constant PAYMASTER_ACCEPTANCE_BUDGET = PRE_RELAYED_CALL_GAS_LIMIT + FORWARDER_HUB_OVERHEAD;
+  uint256 public constant CALLDATA_SIZE_LIMIT = 10500;
+
+  // This is being pulled in to to satisfy the IPaymaster interface.
+  // TODO: I'd like to remove this and depend on the Diamond's implementation.
+  /// @inheritdoc IERC165
+  function supportsInterface(bytes4 _interfaceId) public view override(IERC165) returns (bool) {
+    LibDiamond.DiamondStorage storage ds = LibDiamond.diamondStorage();
+    return ds.supportedInterfaces[_interfaceId];
+  }
+
+  /// @inheritdoc IPaymaster
+  function getRelayHub() public view override(IPaymaster) returns (address) {
+    return address(relayHub);
+  }
+
+  /// @inheritdoc IPaymaster
+  function getTrustedForwarder() public view override(IPaymaster) returns (address) {
+    return trustedForwarder;
+  }
+
+  /// @inheritdoc IPaymaster
+  function getGasAndDataLimits() public pure override(IPaymaster) returns (IPaymaster.GasAndDataLimits memory limits) {
+    return
+      IPaymaster.GasAndDataLimits(
+        PAYMASTER_ACCEPTANCE_BUDGET,
+        PRE_RELAYED_CALL_GAS_LIMIT,
+        POST_RELAYED_CALL_GAS_LIMIT,
+        CALLDATA_SIZE_LIMIT
+      );
+  }
+
+  /// @inheritdoc IPaymaster
+  function preRelayedCall(
+    GsnTypes.RelayRequest calldata relayRequest,
+    bytes calldata /* signature */,
+    bytes calldata approvalData,
+    uint256 /* maxPossibleGas */
+  ) external view override(IPaymaster) onlyRelayHub returns (bytes memory, bool) {
+    if (getTrustedForwarder() != relayRequest.relayData.forwarder)
+      revert IPaymasterErrors.ForwarderNotTrusted(relayRequest.relayData.forwarder);
+
+    // SEE: https://github.com/opengsn/gsn/blob/v3.0.0-beta.10/packages/contracts/src/utils/GsnEip712Library.sol#L59
+    GsnEip712Library.verifyForwarderTrusted(relayRequest);
+
+    if (relayRequest.request.value != 0) revert IPaymasterErrors.ValueTransferNotSupported();
+    if (relayRequest.relayData.paymasterData.length != 0) revert IPaymasterErrors.InvalidPaymasterDataLength();
+    if (approvalData.length != 0) revert IPaymasterErrors.InvalidApprovalDataLength();
+
+    // /=-@=-@=-@=-@=-@=-@=-@=-@=-@=-@=-@=-@=-@=-@/
+
+    // Check for marketplace membership.
+    LibPaymaster.Data storage s = LibPaymaster.data();
+    if (!AHasMembers(s.marketplace).isMember(relayRequest.request.from))
+      revert ICustomErrors.RequiresMarketplaceMembership(relayRequest.request.from);
```

Comment:
The rest is essentially boilerplate checks/validations etc... I wanted to make this part obvious that this is _our_ logic to whitelist the sending account.

Currently just checking for marketplace membership.